### ISSUE TITLE
Refactor to PostCSS and CSS Modules

### ISSUE THREAD

**Issue by chroder** [deskpro/deskpro-components#97]
Was just chatting with Julien about using PostCSS and CSS Modules instead of SCSS.

Goals / Pros:

* PostCSS is faster than SCSS. (As we know from experience, SCSS becomes very slow and starts to impact the feedback-loop while developing!).
* PostCSS uses more standard syntax if that matters; but more important is it's dynamic. We take the features we want and can add things as necessary (vs kitchen sink SCSS).
* CSS Modules make it cleaner to use styles alongside components, and prevents any naming conflicts. (This also matters for dead-code elimination; webpack can trace which things are used and will be able to eliminate unused styles!).
* Also can be used to eliminate conflicts and bugs caused by cascading styles. E.g. Look at [this for example](https://github.com/react-toolbox/react-toolbox/blob/dev/components/avatar/theme.css#L18) where a reset is applied.
* Interesting themeing capabilities (see the example project below)

Cons:

* We're using SCSS at the moment, so obviously there's going to be some refactor time. But Julien seems to think that a switchover would take a day or two.
* Does this preclude the use of these styles outside of React? Are we OK with that?

---

A big project with similar goals to ours:
https://github.com/react-toolbox/react-toolbox

(Indeed, most component libraries seem to go this route.)

---

Eliminating conflicts is the biggest reason this is so appealing to me, the other stuff is just a bonus. Going this route means it'll be easier to pick-n-choose components without bringing in the entire style; and also means we can be confident we can use the components in nearly any page and in any position and know it won't conflict with local style. That's pretty big!

**Comment by jducro** [comment]
About the cons:
- I don't see anywhere where we would use deskpro-components outside React, maybe in jQuery based apps?
- Refactoring is always a bit time consuming but I think the amount of code here is still reasonable

**Comment by helstern** [comment]
About the cons:
- i think there should be a way to use the styles without react. for instance, semantic-ui can be used with or without react. To me this would mean having the final css as another artifact coming out of the build, alongside the one scoped to components

**Comment by jducro** [comment]
@helstern Semantic-ui does not use post-css and css-modules.
Css-modules means inserting style in js from the css file while generating hash classNames.
The style is then way simpler as it doesn't cascade so much anymore

**Comment by chroder** [comment]
It's worth considering that there are two types of styles. There is styling that is very coupled to the implementation, and styles that are more general page styles.

E.g. a select box built in React will have all sorts of styling that is basically useless outside of the react implementation. It's very unlikely we're going to re-implement widgets/controls in other frameworks because it's just not worth the effort. So styling on those is basically only ever going to be used with the React component.

But there are other styles like containers, headings, form layouts, etc etc that can absolutely be used outside of react.

There are two solutions:

* We find a way to export certain CSS modules with "friendly names" so they can be used in plain html.
* Or, we post-process plain HTML (e.g. [this](https://github.com/css-modules/postcss-modules)) and css modules are used in plain html.

**Comment by helstern** [comment]
@jducro i know, i was using it as an example. the point is as @chroder says above, that you should remain flexible and keep your options open. 

I think exporting with friendly names is the solution and i believe there is a plugin for this, at least for webpack

**Comment by chroder** [comment]
AFAIK you can write styles with `:global .my-rule` to mark it as a global styles.

```
:global .myContainer {
    h1 { ... }
    section { ... }
}

// exports as 
.myContainer h1 { ... }
.myContainer section { ... }
```

So perhaps not such a big issue after all. We can still ship with "page styles" for basic stuff, and keep using css modules for components. Best of both worlds.

**Comment by chroder** [comment]
Indeed, switching can be done incrementally this way. Wrap all existing SCSS, and it will act like it does already, and we can refactor out SCSS slowly, or just leave legacy styles as-is and only write new styles.

```
// https://github.com/deskpro/deskpro-components/blob/master/src/styles/main.scss
:global {
    @import ...
    ....
}
```

**Comment by headzoo** [comment]
> Does this preclude the use of these styles outside of React? Are we OK with that?

As it stands, the deskpro-components CSS isn't really designed to be used outside of the library. The styles are essentially "internal" to the components. The fact that _some_ of the CSS could be used outside of the React components is accidental. So we've basically been OK with that from the start.

> i think there should be a way to use the styles without react.

@helstern This wouldn't be possible even if we wanted to. Semantic UI was designed from the start to be a CSS framework, while deskpro-components was designed from the start to be a React components library. Had our intention been to create reusable CSS, we would have first created a CSS framework called deskpro-ui, and _then_ created deskpro-components around it, which is the approach taken by Semantic UI React, which was created 2 years after Semantic UI.

The CSS from dp-components is very much coupled to the HTML structure created by the components, and that HTML structure wasn't designed to be "human friendly." Trying to recreate the components outside of React in order to reuse the CSS would be quite a chore. Even the simple stuff like containers, headings, etc, as @chroder mentioned, weren't really designed from the start to have human friendly/readable HTML structures.

I'm not saying it wouldn't be possible to create a reusable CSS library, but it would have taken a lot more careful planning and many more months to complete, and deskpro-components wouldn't have been nearly as powerful.

**Comment by chroder** [comment]
> Even the simple stuff like containers, headings, etc, as @chroder mentioned, weren't really designed from the start to have human friendly/readable HTML structures.

Mmm not sure I agree with this. The things where this matters are perfectly usable outside of react right now. I don't know if it matters if they weren't "designed to be friendly" really; they're just so simple that they are usable by accident as you say.

* https://github.com/deskpro/deskpro-components/blob/master/src/Components/Buttons/Button.jsx#L96
* https://github.com/deskpro/deskpro-components/blob/master/src/Components/Cards/Card.jsx#L11
* https://github.com/deskpro/deskpro-components/blob/master/src/Components/Common/Heading.jsx#L38 
...

Don't think we particularly want or care that e.g. accordion can't be used without react. But seems like these "simple" building blocks should be usable outside of the framework.

The use case is someone using jQuery to build an app. They wont be able to use any cool widgets, but they should still get basic styling for a table, or a button, or a hr, a numbered list, whatever. It's tedious sure, their html might be full of ugly long classnames, but not sure that matters much.

**Comment by chroder** [comment]
I guess there's a question if those sorts of styles belong in `deskpro-components` anyway. E.g. maybe we should just consider that out-of-scope and leave it to the user to decide where to get the base styles. E.g. bootstrap or foundation, maybe we even help by shipping with a variables list where our colours are assigned to those colours.

**Comment by headzoo** [comment]
> The things where this matters are perfectly usable outside of react right now. 

I know they are. I'm using those components to illustrate a point to @helstern that deskpro-components and semantic-ui have very different goals, so the two projects aren't comparable. I'm making that point because he wasn't involved with deskpro-components, and I don't know how much he knows about the project, and I want to make sure we're on the same page.

Had I felt the need to use a slew of HTML elements to create an amazing button in deskpro-component, I would have done that without thinking twice about it. It's total happenstance that _any_ of the CSS is reusable outside of the components. In other words, no one should be thinking of deskpro-components as a CSS library like semantic-ui because it most certainly isn't one.

> I guess there's a question if those sorts of styles belong in deskpro-components anyway.

It's definitely something to think about, because writing deskpro-components to have reusable CSS is going to require a shift in thinking. Like I mentioned in Slack, it probably makes sense to extract the reusable bits of CSS into it's own library which can be used across the whole DeskPRO ecosystem, e.g. the colours and other variables. Which is something we agree on.
